topic-controller/src/main/java/io/strimzi/controller/topic/TopicMetadataHandler.java

```diff
@@ -0,0 +1,98 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements. See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License. You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package io.strimzi.controller.topic;
+
+import io.vertx.core.AsyncResult;
+import io.vertx.core.Handler;
+import io.vertx.core.Vertx;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.util.concurrent.TimeUnit;
+
+/**
+ * Represents an handler for getting Kafka topic metadata with built-in retry
+ * mechanism with backoff
+ */
+public abstract class TopicMetadataHandler implements Handler<AsyncResult<TopicMetadata>> {
+
+    private static final Logger log = LoggerFactory.getLogger(TopicMetadataHandler.class);
+
+    private final BackOff backOff;
+
+    private final Vertx vertx;
+    private final Kafka kafka;
+    private final TopicName topicName;
+
+    /**
+     * Constructor
+     *
+     * @param vertx Vert.x instance to use for retrying mechanism
+     * @param kafka Kafka client for getting topic metadata
+     * @param topicName topic name for which to get metadata
+     * @param backOff   backoff information to use for retrying
+     */
+    TopicMetadataHandler(Vertx vertx, Kafka kafka, TopicName topicName, BackOff backOff) {
+        this.vertx = vertx;
+        this.kafka = kafka;
+        this.topicName = topicName;
+        this.backOff = backOff;
+    }
+
+    /**
+     * Constructor
+     *
+     * @param vertx Vert.x instance to use for retrying mechanism
+     * @param kafka Kafka client for getting topic metadata
+     * @param topicName topic name for which to get metadata
+     */
+    TopicMetadataHandler(Vertx vertx, Kafka kafka, TopicName topicName) {
+        this(vertx, kafka, topicName, new BackOff());
+    }
+
+    /**
+     * Provides the retry mechanism
+     */
+    protected void retry() {
```

Comment:
I think we should document a little more how the retry works:

"Schedules this handler to execute again after a delay defined by the {@code BackOff}. Calls {@link #onMaxAttemptsExceeded()} if the backoff has reached its permitted number of retries."

That makes the contract much clearer imho.

Reply:
Agree. Thanks !